File: ci/certificate-downloader.py
#!/usr/bin/env python3

from requests import Session
import json

s = Session()

import logging

logger = logging.getLogger(__name__)

URL_ACC = "https://dgca-verifier-service-eu-acc.cfapps.eu10.hana.ondemand.com/"
URL_PRD = "https://get.dgc.gov.it/v1/dgc/signercertificate/"

STATUS_ENDPOINT_ACC = "signercertificateStatus"
UPDATE_ENDPOINT_ACC = "signercertificateUpdate"

STATUS_ENDPOINT = "status"
UPDATE_ENDPOINT = "update"

headers = {
    'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36',
    'x-resume-token' : '0'
}

kid_list = s.get(URL_PRD + STATUS_ENDPOINT, headers=headers).json()

certificate_list = {}

for kid in kid_list:
	response = s.get(URL_PRD + UPDATE_ENDPOINT, headers=headers)
	if kid != response.headers["x-kid"]:
		logger.warning(f"Kid doesn't match: {kid}")
	certificate_list[kid] = response.text
	headers['x-resume-token'] = response.headers['x-resume-token']
	logger.info("Current token: " + headers['x-resume-token'])

with open("certificate_list.json", "w") as f:
	f.write(json.dumps(certificate_list))

File: ci/certificate-downloader.js
const fetch = require('node-fetch')
const fs = require('fs/promises')

const BASE_URL = 'https://get.dgc.gov.it/v1/dgc/signercertificate'
const STATUS_ENDPOINT = 'status'
const UPDATE_ENDPOINT = 'update'

const headers = {
  'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.93 Safari/537.36',
  'x-resume-token': '0'
}

const fetchCertificates = async () => {
  console.info('Fetching kid list...')
  const kidList = await fetch(`${BASE_URL}/${STATUS_ENDPOINT}`, { headers })
    .then(response => response.json())
  console.debug(kidList)

  const certificateList = {}

  console.info('Processing kid list...')
  for (const kidId in kidList) {
    const kid = kidList[kidId]

    const response = await fetch(`${BASE_URL}/${UPDATE_ENDPOINT}`, { headers })
    if (response.headers.get('x-kid') !== kid) {
      console.warn(`Kid doesn't match: ${kid}`)
    }

    certificateList[kid] = await response.text()
    headers['x-resume-token'] = response.headers.get('x-resume-token')
    console.info(`Current token: ${headers['x-resume-token']}`)
  }

  const certificateListJson = JSON.stringify(certificateList)

  console.info('Storing certs on ./certificateList.json')
  console.debug(certificateList)
  await fs.writeFile('certificateList.json', certificateListJson)
  console.info('Certs stored successfully')
}

// if __name__ == '__main__'
if (require.main === module) {
  fetchCertificates()
    .then(() => console.log('Certificate download completed'))
    .catch((err) => console.error(err))
}
